### DISCUSSION TITLE
Converting Pytree params to a vector

### DISCUSSION THREAD

**Discussion by dptam** [jax-ml/jax#9176] [Q&A] [answered]
Hi, I was wondering how to convert a pytree of params to form the a single vector with all the parameters flattened, and then mapping this vector back to the pytree of params. 

1. At the moment, I flatten the pytree of params, and then concatenate the list of values. However, to map this single vector back to the pytree of params, I have to first jnp.split this vector into the a list of individual parameter shapes, reshape each parameter to its original size, and then unflatten this list to form a pytree. It seems jnp.split might not be jittable? I was wondering if there was a more efficient or jittable way to do this operation? 
2. Also, it seems from #4085 that jax.tree_map does not necessarily keep the order of keys in the dictionary. This causes the order of parameters to be different if I first map the Pytree of params to do some operation, and then flatten it. Is there a way around this at the moment? 

Thanks for your help.

**Comment by PhilipVinc** [comment, marked as answer]
[`jax.flatten_util.tree_ravel`](https://jax.readthedocs.io/en/latest/_autosummary/jax.flatten_util.ravel_pytree.html?highlight=ravel#jax.flatten_util.ravel_pytree) does exactly this. It returns the flattened vector + a function to unravel (or unflatten) the vector.

**Reply by dptam** [reply]
Thanks for your help. This is exactly what I'm looking for.

**Reply by ddrous** [reply]
@PhilipVinc This function doesn't appear to exist in JAX anymore. It is super valuable. Is there a replacement ?

**Reply by PhilipVinc** [reply]
It's called ravel_pytrre those days
